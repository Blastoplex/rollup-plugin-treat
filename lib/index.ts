import type { OutputChunk, Plugin, PluginContext } from 'rollup';
import { rollup } from 'rollup';
import type { FilterPattern } from '@rollup/pluginutils';
import { createFilter, dataToEsm } from '@rollup/pluginutils';
import crypto from 'crypto';
import path from 'path';
import type { WebpackTreat as TreatStoreInterface } from 'treat/lib/types/types';
import { TreatTheme } from 'treat/lib/types/types';
import processCss from 'treat/webpack-plugin/processCss';
import eval from 'eval';

import debugLib from 'debug';
import { ThemeOrAny } from 'treat/theme';

const debug = debugLib('rollupPluginTreat');

const THEMED = Symbol('themed treat style');
const LOCAL = Symbol('local treat style');

const treatStoreCache = new WeakMap<PluginContext, TreatStore>(); // TODO: Type me

interface FilterOptions {
	include?: FilterPattern;
	exclude?: FilterPattern;
}

interface Options extends FilterOptions {
	localIdentName?: string;
	themeIdentName?: (() => string) | string;
	outputCSS?: boolean;
}

const isWatchMode = process.env.ROLLUP_WATCH !== 'true'; // TODO: Confirm this is how we check?

const defaultOptions = (options: Partial<Options>): Required<Options> => {
	const {
		exclude = null,
		outputCSS = true,
		include = /\.treat\.(ts|js)$/,
		localIdentName = isWatchMode
			? '[name]-[local]_[hash:base64:5]'
			: '[hash:base64:5]',
		themeIdentName = isWatchMode
			? '_[name]-[local]_[hash:base64:4]'
			: '[hash:base64:4]',
	} = options;

	return {
		exclude,
		localIdentName,
		themeIdentName,
		include,
		outputCSS,
	};
};

export const rollupPluginTreat = (passedOptions: Options): Plugin => {
	const options = defaultOptions(passedOptions);
	const shouldProcessAsTreat = createFilter(options.include, options.exclude);

	return {
		name: 'rollupPluginTreat',

		async transform(code: string, id) {
			if (!shouldProcessAsTreat(id)) return null;

			if (this.cache.has(id)) return this.cache.get(id);

			const sourceCompiled = await childCompile(id);

			let __treat_store__ = treatStoreCache.get(this);

			if (!__treat_store__) {
				__treat_store__ = new TreatStore(options, code, id);
				treatStoreCache.set(this, __treat_store__);
			}

			let result;
			try {
				result = eval(
					`
					require('treat/lib/commonjs/webpackTreat').setWebpackTreat(__treat_store__);
					${sourceCompiled.code}
				`,
					id,
					{
						console,
						__treat_store__,
					},
					true,
				);
			} catch (e) {
				throw e;
			}

			if (options.outputCSS) {
				const cssValue = await processCss(
					{
						'.test': {
							color: 'red',
						},
					},
					{
						browsers: [],
						minify: false,
						from: id,
					},
				);

				this.emitFile({
					type: 'asset',
					source: cssValue,
				});
			}

			return dataToEsm(result, {
				namedExports: true,
				preferConst: true,
			});
		},
	};
};

const childCompile = async (id: string): Promise<OutputChunk> => {
	const fileBundle = await rollup({
		input: id,
	});

	const { output } = await fileBundle.generate({
		format: 'cjs',
	});

	if (output.length > 1)
		throw new Error(
			"Didnt expect child compiler to produce more than 1 file. Perhaps you're using async imports?",
		);

	return output[0];
};

class TreatStore implements TreatStoreInterface {
	constructor(
		private options: ReturnType<typeof defaultOptions>,
		private sourceCode: string,
		private id: string,
	) {}

	themeIdentFn = processIdent(
		this.options.themeIdentName,
		this.sourceCode,
		this.id,
	);
	localIdentFn = processIdent(
		this.options.localIdentName,
		this.sourceCode,
		this.id,
	);

	themes: Array<TreatTheme<ThemeOrAny>> = [];

	styles = new Set();

	addLocalCss(css) {}

	addThemedCss(themeRef, css) {}

	getIdentName(local, scopeId, theme) {
		return theme !== undefined
			? this.themeIdentFn(local, scopeId, theme)
			: this.localIdentFn(local, scopeId);
	}

	addTheme(theme) {
		this.themes.push(theme);
	}

	getThemes() {
		return this.themes;
	}
}

const processIdent = (identFn, sourceCode: string, id: string) => {
	return (localName: string, scopeId: string, theme?: string): string => {
		const identName =
			typeof identFn === 'function' ? identFn(theme) : identFn;

		return identName
			.replace(
				/\[(?:([^:\]]+):)?(?:hash|contenthash)(?::([a-z]+\d*))?(?::(\d+))?\]/gi,
				(_all, hashType = 'md5', digestType, maxLength = 10) =>
					crypto
						.createHash(hashType)
						.update(`${sourceCode}${scopeId}`)
						.digest(digestType)
						.slice(0, maxLength),
			)
			.replace(/\[local]/gi, localName)
			.replace(/\[name]/gi, path.basename(id, path.extname(id)))
			.replace('treat-', '')
			.replace(new RegExp('[^a-zA-Z0-9\\-_\u00A0-\uFFFF]', 'g'), '-')
			.replace(/^((-?[0-9])|--)/, '_$1');
	};
};
