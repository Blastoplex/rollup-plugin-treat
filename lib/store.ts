import type { ThemeOrAny } from 'treat/theme';
import type { ThemeRef, TreatTheme } from 'treat/lib/types/types';
import { d } from './util';

export const enum STYLE_TYPE {
	LOCAL = 'local',
	THEME = 'theme',
}

export type StyleItem =
	| { type: STYLE_TYPE.LOCAL; styles: object }
	| { type: STYLE_TYPE.THEME; styles: object; themeRef: ThemeRef };

export const store = () => {
	const themes = new Map<string, { theme: TreatTheme<ThemeOrAny> }>();
	const styleResources = new Map<string, Array<StyleItem>>();

	const getThemes = (): Array<TreatTheme<ThemeOrAny>> => {
		return Array.from(themes.values()).map(({ theme }) => theme);
	};

	const addTheme = (theme: TreatTheme<ThemeOrAny>) => {
		if (!themes.has(theme.themeRef))
			d('discovered theme %s', theme.themeRef);

		themes.set(theme.themeRef, {
			theme,
		});
	};

	const addStyleResource = (id: string, styleItems: Set<StyleItem>) => {
		const currentResource = styleResources.get(id) ?? [];

		styleResources.set(
			id,
			[...currentResource, ...styleItems].sort((a, b) => {
				if (a.type === b.type) {
					return 0;
				} else if (a.type === STYLE_TYPE.LOCAL) {
					return -1;
				} else {
					return 1;
				}
			}),
		);
	};

	const getStyleResource = (id: string) => styleResources.get(id);

	return {
		getThemes,
		addTheme,
		addStyleResource,
		getStyleResource,
	};
};
export type StoreType = ReturnType<typeof store>;
