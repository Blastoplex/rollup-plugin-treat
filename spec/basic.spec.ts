import { rollup } from 'rollup';
import fs from 'fs';
import { resolve } from 'path';

import { rollupPluginTreat } from '../lib';

const fixtureFolder = resolve(__dirname, './fixtures/');

const fixtureInputs = fs.readdirSync(fixtureFolder, {
	withFileTypes: false,
});

// @ts-ignore
test.each(fixtureInputs as any)('it should snapshot match %s', async (testFolder: string) => {
	const bundle = await rollup({
		input: resolve(fixtureFolder, testFolder, './main.js'),
		plugins: [
			rollupPluginTreat({}),
		],
	});

	const { output } = await bundle.generate({
		format: 'cjs',
	});

	output.forEach(item => {
		if (item.type === 'chunk') {
			expect(item.code).toMatchSnapshot(item.name);
		} else if (item.type === 'asset') {
			expect(item.source).toMatchSnapshot(item.fileName);
		}
	});
});
