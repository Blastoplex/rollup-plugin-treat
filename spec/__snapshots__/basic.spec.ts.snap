// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should snapshot match async-modules: a 1`] = `
"import { resolveStyles } from 'treat';

const a = \\"a-style_hnUgi\\";
var a_treat = {
	a: a
};

var styleRefs = /*#__PURE__*/Object.freeze({
	__proto__: null,
	a: a,
	'default': a_treat
});

console.log(
	resolveStyles(null, styleRefs),
);
"
`;

exports[`it should snapshot match async-modules: bundle.css 1`] = `".a-style_hnUgi{color:red}"`;

exports[`it should snapshot match async-modules: main 1`] = `
"import('./a-4eec1be4.js')
	.then(() => {

	});
"
`;

exports[`it should snapshot match basic-no-theme: bundle.css 1`] = `".main-style_OsUq7{color:red}"`;

exports[`it should snapshot match basic-no-theme: main 1`] = `
"import { resolveStyles } from 'treat';

const myClass = \\"main-style_OsUq7\\";
var main_treat = {
	myClass: myClass
};

var styleRefs = /*#__PURE__*/Object.freeze({
	__proto__: null,
	myClass: myClass,
	'default': main_treat
});

console.log(
	resolveStyles(null, styleRefs),
);
"
`;

exports[`it should snapshot match basic-with-theme: bundle.css 1`] = `".main-style_fzVq4{color:pink}.main-style_wvq38_theme-theme_hBG0{color:red}"`;

exports[`it should snapshot match basic-with-theme: main 1`] = `
"import { resolveStyles } from 'treat';

const myTheme = \\"_theme-theme_hBG0\\";

const plainClass = \\"main-style_fzVq4\\";
const themedClass = \\"$main-style_wvq38\\";
var main_treat = {
	plainClass: plainClass,
	themedClass: themedClass
};

var styleRefs = /*#__PURE__*/Object.freeze({
	__proto__: null,
	plainClass: plainClass,
	themedClass: themedClass,
	'default': main_treat
});

console.log(
	resolveStyles(myTheme, styleRefs),
);
"
`;

exports[`it should snapshot match module-order: bundle.css 1`] = `".a-style_M-kUM{color:red}.b-style_n3t8V{color:#00f}"`;

exports[`it should snapshot match module-order: main 1`] = `
"import { resolveStyles } from 'treat';

const a = \\"a-style_M-kUM\\";
var a_treat = {
	a: a
};

var styleRefs = /*#__PURE__*/Object.freeze({
	__proto__: null,
	a: a,
	'default': a_treat
});

const b = \\"b-style_n3t8V\\";
var b_treat = {
	b: b
};

var styleRefs$1 = /*#__PURE__*/Object.freeze({
	__proto__: null,
	b: b,
	'default': b_treat
});

console.log(
	resolveStyles(null, styleRefs$1),
);

console.log(
	resolveStyles(null, styleRefs),
);
"
`;

exports[`it should snapshot match using-node-modules: bundle.css 1`] = `".main-style_K8-5m{color:rgba(255,205,100,.7)}"`;

exports[`it should snapshot match using-node-modules: main 1`] = `
"import { resolveStyles } from 'treat';

const myClass = \\"main-style_K8-5m\\";
var main_treat = {
	myClass: myClass
};

var styleRefs = /*#__PURE__*/Object.freeze({
	__proto__: null,
	myClass: myClass,
	'default': main_treat
});

console.log(
	resolveStyles(null, styleRefs),
);
"
`;
